modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildingStarterTask.java

```diff
@@ -0,0 +1,321 @@
+
+                            if (cause instanceof PrimaryReplicaAwaitTimeoutException) {
+                                return awaitPrimaryReplica();
+                            }
+
+                            return CompletableFuture.<ReplicaMeta>failedFuture(cause);
+                        }
+
+                        return completedFuture(replicaMeta);
+                    }).thenCompose(Function.identity());
+        });
+    }
+
+    private CompletableFuture<Void> awaitFinishRwTxsBeforeCatalogVersionOfIndexCreation(LogicalTopologySnapshot logicalTopologySnapshot) {
+        return inBusyLock(() -> {
+            Set<LogicalNode> remainingNodes = ConcurrentHashMap.newKeySet();
+            remainingNodes.addAll(logicalTopologySnapshot.nodes());
+
+            var nodeLeftLogicalTopologyListener = new NodeLeftLogicalTopologyListener(remainingNodes);
+            logicalTopologyService.addEventListener(nodeLeftLogicalTopologyListener);
+
+            CompletableFuture[] futures = remainingNodes.stream()
+                    .map(node -> awaitFinishRwTxsBeforeCatalogVersionOfIndexCreation(node, remainingNodes))
+                    .toArray(CompletableFuture[]::new);
+
+            return allOf(futures)
+                    .whenComplete((unused, throwable) -> logicalTopologyService.removeEventListener(nodeLeftLogicalTopologyListener));
+        });
+    }
+
+    private CompletableFuture<NetworkMessage> awaitFinishRwTxsBeforeCatalogVersionOfIndexCreation(
+            LogicalNode targetNode,
+            Set<LogicalNode> remainingNodes
+    ) {
+        return inBusyLock(() -> {
+            if (!remainingNodes.contains(targetNode)) {
+                // Target node left logical topology, there is no need to do anything.
+                return nullCompletedFuture();
+            }
+
+            IsNodeFinishedRwTransactionsStartedBeforeRequest request = isNodeFinishedRwTransactionsStartedBeforeRequest();
+
+            return clusterService.messagingService().invoke(targetNode, request, INVOKE_MESSAGE_TIMEOUT_MILLS)
+                    .handle((message, throwable) -> {
+                        if (throwable != null) {
+                            Throwable cause = unwrapCause(throwable);
+
+                            if (!(cause instanceof TimeoutException) && !(cause instanceof RecipientLeftException)) {
+                                return CompletableFuture.<NetworkMessage>failedFuture(cause);
+                            }
+                        } else if (((IsNodeFinishedRwTransactionsStartedBeforeResponse) message).finished()) {
+                            remainingNodes.remove(targetNode);
+
+                            return completedFuture(message);
+                        }
+
+                        // Let's try to send the message again, but after a short time.
+                        return inBusyLock(() -> clockWaiter.waitFor(clock.now().addPhysicalTime(RETRY_SEND_MESSAGE_TIMEOUT_MILLS)))
```

Comment:
I'm not sure we should use Clock and ClockWaiter just to schedule a task after X milliseconds. For this, a usual `ScheduledThreadPool` should be enough. There is a common scheduled pool in `ThreadPoolsManager`, for example.

Reply:
try fix it